packages/@lwc/engine-core/src/framework/hooks.ts

```diff
@@ -265,3 +275,17 @@ export function markAsDynamicChildren(children: VNodes) {
 export function hasDynamicChildren(children: VNodes): boolean {
     return FromIteration.has(children);
 }
+
+export function connectedCallbackHook(vnode: VCustomElement, vm: VM) {
+    if (vm.shadowMode === ShadowMode.Synthetic && vm.renderMode === RenderMode.Shadow) {
```

Comment:
I am quite worried that we have to traverse the DOM upward every time a component connects. 

If we take a step back, the goal of preventing the usage of synthetic shadow components within native shadow components is to prevent developers from shooting themselves in the foot. If a synthetic shadow component relies on global styles leaking into it, the component will stop working when used inside a native shadow DOM component.

IMHO, the correct way to approach this is to climb the component tree using the `VM.owner`. This approach should cover 99.9% of the standard LWC use cases (except dynamic component inject via `lwc:dom=manual`). The goal of a guardrail is to cover the most common use cases but don't need to address all the crazy corner cases.

Reply:
I totally agree that this approach is scary perfwise. I partially opened this PR just to start the conversation about it. (Best way to get the right answer on the internet is to propose the wrong answer. 😄)

If we're okay with not covering 100% of usecases, then absolutely, we should just climb the `VM.owner` tree and not worry about `lwc:dom=manual` (which BTW does not apply here because we're talking about a native shadow tree).

OTOH if we're not aiming to cover 100% of usecases, why not restrict the asserts to `process.env.NODE_ENV === 'development'`?

Reply:
@pmdartus I implemented your suggestion. We can still handle synthetic-within-light-within-native, but not dynamic insertion anymore. But that's okay, because [the benchmark](https://git.soma.salesforce.com/nlawson/lwc-barebone/commit/13615775b0c9c6ae790de15769e8c5ec5a54377f) is much improved: ~21ms doing the owner check during a total rendering time of 916ms:

![Screen Shot 2021-06-08 at 2 52 26 PM](https://user-images.githubusercontent.com/283842/121262688-636ad880-c869-11eb-90dd-d78657b27094.png)

Reply:
Micro-optimized it down even further, to ~3ms:

![Screen Shot 2021-06-08 at 2 59 29 PM](https://user-images.githubusercontent.com/283842/121263226-2b17ca00-c86a-11eb-8a29-7a941c114248.png)

Reply:
> OTOH if we're not aiming to cover 100% of usecases, why not restrict the asserts to process.env.NODE_ENV === 'development'?

Yes, I would recommend only enabling this in `development` mode like other guardrails.